plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java

```diff
@@ -0,0 +1,540 @@
+    Object[] ends = fragments.toArray();
+
+    // The list of fragments contains both ends of each sequence message (sender and receiver)
+    // and only one end of each execution (the one where execution ends). This means that we should skip
+    // the receiving end for each message, so that we don't duplicate the generated xtext message.
+    int i = 0;
+    Stack<org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage> messagesToDeactivate = new Stack();
+    while (i < ends.length) {
+      if (ends[i] instanceof MessageEnd) {
+        EObject message = copyMessageFromMsgEnd(ends[i], factory);
+        messagesOrReferences.add(message);
+        // skip the next MessageEnd (the receiving end), as it will generate the same xtext message
+        i = i + 2;
+        if (ends[i] instanceof ExecutionEnd) {
+          // sikp generating deactivate keyword if we have a simple message
+          i = i + 1;
+        } else if (message instanceof org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage) {
+          messagesToDeactivate.push((org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage) message);
+        }
+      } else {
+        EObject participantDeactivateMsg = getParticipantDeactivationMsgFromExecutionEnd(ends[i], factory);
+        messagesOrReferences.add(participantDeactivateMsg);
+
+        if (!messagesToDeactivate.isEmpty()) {
+          org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage currentSequenceMessage = messagesToDeactivate
+              .pop();
+          currentSequenceMessage.setExecution(DslConstants.WITH_EXECUTION);
+        }
+        i = i + 1;
+      }
+    }
+  }
+
+  /**
+   * generates the ParticipantDeactivation message with input from the ExecutionEnd in the Capella diagram
+   * 
+   * @param object
+   *          - this is the ExecutionEnd
+   * @param factory
+   *          - this is the factory to create ParticipantDeactivation type of message
+   * @return - EObject containing the ParticipantDeactivation message
+   */
+  private static EObject getParticipantDeactivationMsgFromExecutionEnd(Object object, DslFactory factory) {
+    ExecutionEnd end = (ExecutionEnd) object;
+    SequenceMessage seqMessage = ExecutionEndExt.getMessage(end);
+    String timelineToDeactivate = seqMessage.getReceivingEnd().getCoveredInstanceRoles().get(0).getName();
+
+    ParticipantDeactivation participantDeactivationMsg = (ParticipantDeactivation) factory
+        .createParticipantDeactivation();
+    participantDeactivationMsg.setName(timelineToDeactivate);
+
+    participantDeactivationMsg.setKeyword(DslConstants.DEACTIVATE);
+    return participantDeactivationMsg;
+  }
+
+  private static EObject copyMessageFromMsgEnd(Object object, DslFactory factory) {
+    EObject seqMessage;
+    MessageEnd end = (MessageEnd) object;
+    SequenceMessage sequenceMessage = end.getMessage();
+
+    switch (sequenceMessage.getKind()) {
+    case ASYNCHRONOUS_CALL:
```

Comment:
There are duplicated block in the switch. Could you merge them?

Reply:
OK.